### DISCUSSION TITLE
Higher Order Polymorphism chapter

### DISCUSSION THREAD

**Discussion by Bohun9** [ppolesiuk/type-systems-notes#101] [Q&A]
1. I observed that type equivalence relation in "Types and Programming Languages" book (451 page) is not parameterized with delta and I don't understand why.

2. I think that gamma should store types of variables and delta should store kinds of type variables, so kind system and type equivalence should be parameterized with only delta and type system with both? I can fix this in notes, but I want to check first if I understand it correctly.

3. Why there is no type lambda abstraction/type application case in Relation Model?

4. Is it true that we only care about type equivalence of kind *? Because I think $\lambda x :: *. x$ and $\lambda y :: *. y$ are not equivalent now.

**Comment by ppolesiuk** [comment]
1. There is some design space here, and Pierce in his TAPL book made a different choice. The equivalence relation in the book is untyped, but the conversion rule (T-Eq) ensures that we convert type to an equivalent well-kinded type. In our approach we build equivalence relation that is well-kinded by design, and therefore, we parametrize it with Δ to have access to kinds of type variables. The environment Δ is used in the premise of β-rule (to ensure, that the argument is well-kinded), and should be also used in the reflexivity rule (reflexivity is partial, and works only for well-kinded types). To be even more correct, we could parametrize the relation with the kind of equated types: the judgment would have the form Δ ⊢τ₁ ≡ τ₂ ∷ κ. However, in our case each type has at most one kind, so this additional parameter is not necessary.
2. Yes, there is an obvious typo in the notes.
3. There is a mess in the interpretation now. The interpretation of arrow, quantifier, and lambda are missing, the interpretation of application contains some arrow on the left-hand-side. It should be fixed. Interpretations of arrow and quantifier should be standard, while the interpretation should be obviously a lambda on a meta-level: ⟦ λ α∷κ.τ ⟧_η = λ R∈⟦κ⟧. ⟦τ⟧η[α↦R].
4. No. We care about the type equivalence on any kind. In the typing rules we only use the type equivalence on kind ∗, but the equivalence is defined as the smallest (well-kinded) congruence containing β-rule. Since it is a congruence, it contain structural rules, and structural rule for application requires equivalence on an arrow kind. By the way, λ α ∷ ∗. α and λ β ∷ ∗. β are equivalent, because they are equal: we always work modulo α-equivalence.

**Reply by Bohun9** [reply]
1. In TAPL, rule (T-Eq) asserts that types have kind * and in our system it can be any kind. Is it important difference?

**Reply by ppolesiuk** [reply]
It can be proven that in our system the conversion rule can be applied to ∗ kind only. First, the type equivalence can equate only types of the same kind. Second, we can derive judgment Δ;Γ⊢e:τ only for τ of kind ∗ (or more precisely, we have Δ⊢τ∷∗).

**Reply by Bohun9** [reply]
You said in the answer to (1) that equivalence relation is well-kinded by design, but I came across some weird case: in $\beta$-rule, kind of $\tau$ is not checked. It looks weird, because we take denotation of this type in task 2, and it can obviously be not defined. I guess each rule should assert kinds of the types it is using, right?

**Reply by ppolesiuk** [reply]
Formally, we defined the denotation for types well-kinded by design (the second approach), so this is not an issue. But, you are right: to ensure that type equivalence relation is well-kinded by design, we have to assert kinds of each (or most of) the types it is using. Such well-kindedness premises obscures a presentation a bit, so it is convenient to assume them by convention, not writing them explicitly.
